### ISSUE TITLE
Recommendations

### ISSUE THREAD

**Issue by codex-zaydek** [qawolf/qawolf#442]
Just a bunch of small recommendations I thought I’d share

1. Currently QA Wolf enforces camel-casing, which might be appealing but I found it surprising:

```
yarn qawolf create http://localhost:3000 my-test
-> myTest
```

2. The docs are good but a little light. This isn’t a bad thing, but it feels _slightly_ unprofessional and unconvincing to prospective users. I imagine users want to feel safe using a new tool, and having a little bit beefier docs can help with that.

I also recommend adding explicit code blocks for Yarn. It wasn’t obvious to me that `npx run qawolf` and `yarn qawolf` are the same. Are they?

3. The text serialization for typing is weird. I get it, it’s even a nice idea, but in the current form it didn’t feel right.

For example:

```
↓ShiftLeft↓Digit3↑Digit3↑ShiftLeft↓ShiftLeft↑ShiftLeft↓Space↑Space↓ShiftLeft↓KeyH↑ShiftLeft↑KeyH↓KeyE↑KeyE↓KeyL↑KeyL↓KeyL↑KeyL↓KeyO↓Comma↑KeyO↓Space↑Comma↓KeyW↓KeyO↑Space↑KeyW↑KeyO↓KeyR↓KeyL↑KeyR↓KeyD↑KeyL↑KeyD↓ShiftLeft↓Digit1↑ShiftLeft↑Digit1↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↓Backspace↑Backspace↓ShiftLeft↓Digit3↑Digit3↓Space↑Space↓KeyH↑KeyH↑ShiftLeft↓KeyE↑KeyE↓KeyL↑KeyL↓KeyL↑KeyL↓KeyO↑KeyO↓Comma↓Space↑Comma↓KeyW↑Space↓KeyO↑KeyW↓KeyR↑KeyO↓KeyL↑KeyR↓KeyD↑KeyL↑KeyD↓ShiftLeft↓Digit1↑Digit1↑ShiftLeft
```

This is the output I got when I typed, backspaced, and typed `# Hello, world!` again. There are a couple of things I want to expand on with why I think the current approach is wrong:

I totally get the modifier thing, but I think it makes more sense to separate the non-typing logic. So for example, with the `input` event handler for editable elements like `<textarea>`, etc., most events are `e.inputType === "inserText"` or event `e.inputType === "insertCompositionText"`, but backspace, delete (forward backspace) are different, e.g. `e.inputType === "deleteContentBackward"`, etc. There’s a bunch of them: https://www.w3.org/TR/input-events-1/#interface-InputEvent-Attributes.

You don’t need to support these but it can help you reason about which kinds of events should be broken up. Finally, if you do stick with serialization, I **strongly** recommend you also add a comment with a more human-readable version of what the user typed, in my case `# Hello, world!`.

In my current Playwright tests, I’m doing the following:

```js
test("can type and delete characters", async () => {
	await ppt.clear(page)
	await ppt.type(page, "Hello, world! 😀\n\nHello, world! 😀\n\nHello, world! 😀")
	let data = await ppt.innerText(page)
	expect(data).toBe("Hello, world! 😀\n\nHello, world! 😀\n\nHello, world! 😀")
	for (let index = 0; index < 52; index++) {
		await ppt.backspaceChar(page)
	}
	data = await ppt.innerText(page)
	expect(data).toBe("")
})
```

It’s really easy for me to reason about, and could even by simpler/shorter. If this were serialized, I don’t know. I’m not sure that’s the right abstraction here or even the only one.

4. If you want to beat the other tools on the market, like Cypress, even Playwright, etc. I think it will help the most to create a simple and consistent an API as possible, and emphasize how QA Wolf easily integrates with your existing apps. One of the biggest turn offs about Cypress for me was the whole _learn how we do things and then attempt to copy us_ kind of approach. Learning Cypress felt **huge** and I just don’t have the time. I also don’t want to write my tests in anything but a Jest environment because that’s what I understand (coming from CRA), so the more you can emphasize you have a solution that integrates painlessly with apps that are already working today (not future apps that don’t exist), I think that will help a lot.

I really love the idea that you might be able to abstract user events down to as simple events as possible like `click`, `type`, etc. and hide all the implementation details (like selectors) away. I think this is a really powerful thing you can do, and the better you do it, the better precedent you set for all tools that come after you, or are even competing with you. Because you are new, you can actually make some of the most impact by trying new approaches that the incumbent software can’t easily pivot to right now.

5. Wanted to mention that Firefox opens in the background (also not fullscreen), and Safari opens with poor resolution. This is how I fixed that with Playwright:

```js
// Opens a new page from a browser string and a URL.
export async function openPage(browserStr, url) {
	const browserType = browserTypes[browserStr]
	const args = []
	if (browserType === Chrome) {
		args.push("--window-size=1440,900")
	} else if (browserType === Firefox) {
		args.push("-width=1440", "-height=900")
	}
	const config = {
		headless: process.env.HEADLESS === "true" || false,
		args,
	}
	const browser = await browserType.launch(config)
	const context = await browser.newContext()
	const page = await context.newPage(url)
	if (!config.headless && browserType === Firefox) {
		execSync("osascript -e 'activate application \"Nightly\"'")
	}
	await page.setViewport({ width: 1440, height: 900 }) // , deviceScaleFactor: 2 })
	await page.addScriptTag({ path: "./src/components/Editor/__tests/innerText.js" })
	await page.waitFor(1e3, options)
	page.on("pageerror", error => expect(error).toBeNull())
	return [page, () => browser.close()]
}
```

At one point I needed to add `deviceScaleFactor: 2` to make Safari look sharp (I can’t remember if I need that anymore or not). Also, other lines like `args.push("-width=1440", "-height=900")` and `execSync("osascript -e 'activate application \"Nightly\"'")` were needed to get Firefox fullscreen and on top (of the other windows). I’m cheating by using AppleScript to force the window on top, but you get the idea. Small things like this that you can do will make your users feel more _safe_ using your tool because it meets and or exceeds their expectations.

Also, one of the things that _really_ bothers me about Playwright is that they name the browsers `chromium`, `firefox`, and `webkit`. What I would do is this: support both vendor-like names and brand-like names.

So for example:

- `chrome` and `chromium`
- `firefox` and `gecko`
- `safari` and `webkit`

This is how developers think, either as an end-user or as a developer, so I would make these kinds of things very approachable to a user who’s never worked with Puppeteer or Playwright. Details like that can really help your users feel like they’re investing their time well because even the little things are being addressed.

---

That’s all for now — I hope this is actually useful. Feel free to ping me whenever. :)

**Comment by flaurida** [comment]
This is very helpful and I agree with all of this. Thanks so much 🎉 😄 We look forward to implementing these!

**Comment by jperl** [comment]
Cannot thank you enough for this feedback. We will work it all into our immediate roadmap. Relates to #432, #439, #440, #441, #443

**Comment by codex-zaydek** [comment]
When you have another significant release I’d be happy to follow-up and provide critical notes again with fresh eyes.

**Comment by jperl** [comment]
Thank you we will definitely take you up on that 🙏

**Comment by codex-zaydek** [comment]
I’m learning Cypress now and it looks like this is how the deal with command serialization, in addition to separate commands:

https://docs.cypress.io/api/commands/type.html#Arguments

Like `{backspace}`. This includes modifiers like `{meta}`, for example.

**Comment by codex-zaydek** [comment]
This is pretty expressive because I can intuit stuff like this:

```js
cy.get("#editor").type(`# Hello, world!${"{backspace}".repeat(15)}`)
```

**Comment by jperl** [comment]
That's awesome, we will do it this way 👍

**Comment by jperl** [comment]
Just following up on this:

1. Fixed, no longer enforce camel casing.
2. Prioritizing docs improvements next.
3. Fixed, we removed this text serialization and switched to vanilla Playwright.
4. We slimmed down the API and switched to vanilla Playwright.
5. I believe these launch issues are fixed by Playwright in their next release.

Re: naming, I am torn on this, since we want to match Playwright as much as possible. I think we will keep as is for now, but reconsider in the future.

Thanks again for all of your feedback. I believe we incorporated most of it, and if there are still issues let's prioritize those. 🙏
